FIX use unique values of y_true and y_pred in plot_confusion_matrix instead of estimator.classes_

kyouma opened the pull request:
Although the docstring and the API guide of **sklearn.metrics.plot_confusion_matrix()** say about the **labels** argument the following: "_If 'None' is given, those that appear at least once in 'y_true' or 'y_pred' are used in sorted order_", the **estimator.classes_** field was used.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
This change fixes errors when **y_true** and **y_pred** doesn't have some values from **estimator.classes_**.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

kyouma committed:
Fix for the "labels=None" behaviour

Although the docstring and the API guide say "If 'None' is given, those that appear at least once in `y_true` or `y_pred` are used in sorted order", the "estimator.classes_" field was used.

kyouma committed:
FIX for plot_confusion_matrix() when 'labels' is equal to 'None'

Although the docstring and the API guide say "If 'None' is given, those that appear at least once in 'y_true' or 'y_pred' are used in sorted order", the 'estimator.classes_' field was used.

thomasjpfan reviewed (commented):
Thank you for the PR!

Please add a [non-regression test](https://en.wikipedia.org/wiki/Non-regression_testing) that would fail at master but pass in this PR.

kyouma committed:
Added a test for plot_confusion_matrix() behaviour with unseen labels

A test for plot_confusion_matrix() behaviour when 'labels=None' and the dataset with true labels contains labels previously unseen by the classifier (and therefore not present in its 'classes_') attribute. According to the function description, it must create a union of the predicted labels and the true labels.

kyouma committed:
Update 'test_error_on_a_dataset_with_unseen_labels()'

An update to the 'test_error_on_a_dataset_with_unseen_labels()' function to fix 'E501 line too long' errors.

kyouma committed:
Fixes for test_error_on_a_dataset_with_unseen_labels()

kyouma committed:
Fix for test_error_on_a_dataset_with_unseen_labels()

kyouma commented:
> 
> 
> Thank you for the PR!
> 
> Please add a [non-regression test](https://en.wikipedia.org/wiki/Non-regression_testing) that would fail at master but pass in this PR.

Thank you for the review, @thomasjpfan! This is my first pull request, I will try to do my best to implement and prepare everything correctly.

I have added the test **test_error_on_a_dataset_with_unseen_labels()** that checks tick labels of the confusion matrix plot.
In iPython console **matplotlib** doesn't throw exceptions on this test, so I had to add this check. And in Jupyter Notebook the very call of **plot_confusion_matrix()** would raise the exception _ValueError("The number of FixedLocator locations (...), usually from a call to set_ticks, does not match the number of ticklabels (...).")_.
The updated **plot_confusion_matrix()** function is intended to pass this test.

thomasjpfan commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -261,7 +262,20 @@ def plot_confusion_matrix(estimator, X, y_true, *, labels=None,
+                # If `y_true` was provided to the classifier as strings,
+                # `y_pred` given by the classifier will also be encoded with
+                # strings. So we raise a meaningful error
+                raise TypeError(
+                    f"Labels in y_true and y_pred should be of the same type. "
+                    f"Got y_true={np.unique(y_true)} and "
+                    f"y_pred={np.unique(y_pred)}. Make sure that the "
+                    f"predictions provided by the classifier coincides with "
+                    f"the true labels."
```
Do we have a test to make sure this error is raised?

kyouma replied:
I have removed the Try-Except wrapping as the function _confusion_matrix_(), which is used above to get the matrix itself, contains the same _unique_labels_() call that was wrapped by the Try-Except block, and the function _unique_labels_() raises an exception with a description when the true and predicted labels have different types. So if the execution arrives at the line, it will not make any problems.

thomasjpfan commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -314,3 +315,18 @@ def test_default_labels(pyplot, display_labels, expected_labels):
+                                 labels=None, display_labels=None)
+
+    disp_labels = set([tick.get_text() for tick in disp.ax_.get_xticklabels()])
+    expected_labels = unique_labels(y, fitted_clf.predict(X))
```
In this case, we can list the labels:

```py
    display_labels = [tick.get_text() for tick in disp.ax_.get_xticklabels()]
    expected_labels = [f'{i}' for range(6)]
    assert_array_equal(expected_labels, display_labels)
```

kyouma replied:
Thank you, I have replaced these lines and the assertion check with your code.

kyouma committed:
Remove unnecessary try-catch

This Try-Catch is not necessary, as the same unique_labels() function is called inside confusion_matrix() above and raises an exception with a description if the types of true and predicted labels differ.

kyouma committed:
Improvement of the assertion check in test_error_on_a_dataset_with_unseen_labels()

kyouma committed:
Improvement of the assertion check in test_error_on_a_dataset_with_unseen_labels()

thomasjpfan commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -314,3 +315,16 @@ def test_default_labels(pyplot, display_labels, expected_labels):
+
+    # get values that are not present in 'fitted_clf.classes_'
+    y = y + 1
+    disp = plot_confusion_matrix(fitted_clf, X, y,
+                                 labels=None, display_labels=None)
```
```suggestion
    disp = plot_confusion_matrix(fitted_clf, X, y)
```

kyouma replied:
Thank you very much for your attention and help!
I wanted to make sure that the _plot_confusion_matrix_() function is called with _labels=None_ and _display_labels=None_ even if the default behaviour is changed in future. Should I still remove these arguments?

thomasjpfan replied:
In this case, we want to test the normal use case, where a user does not provided the defaults.

If the defaults change, we would update this test.

thomasjpfan commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -314,3 +315,16 @@ def test_default_labels(pyplot, display_labels, expected_labels):
     assert_array_equal(y_ticks, expected_labels)
+
+
+def test_error_on_a_dataset_with_unseen_labels(pyplot, fitted_clf, data):
```
We may need to wrap this to be <= 79:

```suggestion
def test_error_on_a_dataset_with_unseen_labels(pyplot, fitted_clf, data, n_classes):
```

kyouma replied:
Thank you.

thomasjpfan reviewed (commented):
Please add an entry to the change log at `doc/whats_new/v0.24.rst` with tag |Fix|. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

kyouma committed:
Improvement of the assertion check in test_error_on_a_dataset_with_unseen_labels()

- Replaced the assertion check
- Removed the unused import

kyouma committed:
Improvements and fixes for test_error_on_a_dataset_with_unseen_labels()

kyouma committed:
Mentioned the fix for plot_confusion_matrix()

Mentioned the PR #18405.

kyouma committed:
Improvement for test_error_on_a_dataset_with_unseen_labels()

The `labels` and `display_labels` parameters have been set to thier default values.

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

kyouma commented:
Thank you very much, I have implemented your suggestions and corrections. I have also added the |Fix| entry to `doc/whats_new/v0.24.rst`.
There is one checktest that fails: `test_pandas_dependency_message()` raises `MemoryError` (probably at `datasets/tests/test_kddcup99.py`), although I have not changed any of the dataset files.

thomasjpfan commented on doc/whats_new/v0.24.rst:
```diff
@@ -398,6 +398,12 @@ Changelog
   ``labels`` parameter.
   :pr:`17935` by :user:`Cary Goltermann <Ultramann>`.
 
+- |Fix| bug in :func:`metrics.plot_confusion_matrix` where error occurs when
```
May need to rewrap:

```suggestion
- |Fix| Fixed bug in :func:`metrics.plot_confusion_matrix` where error occurs when
```

thomasjpfan commented on doc/whats_new/v0.24.rst:
```diff
@@ -398,6 +398,12 @@ Changelog
   :pr:`17935` by :user:`Cary Goltermann <Ultramann>`.
 
+- |Fix| bug in :func:`metrics.plot_confusion_matrix` where error occurs when
+  ``y_true`` contains labels that were not previously seen by the classifier
+  while the ``labels`` and ``display_labels`` parameters are set to `None`.
```
```suggestion
  `y_true` contains labels that were not previously seen by the classifier
  while the `labels` and `display_labels` parameters are set to `None`.
```

thomasjpfan commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -314,3 +314,16 @@ def test_default_labels(pyplot, display_labels, expected_labels):
+                                               n_classes):
+    X, y = data
+
+    # get values that are not present in 'fitted_clf.classes_'
```
```suggestion
    # use values that are not present in 'fitted_clf.classes_'
```

thomasjpfan reviewed (approved):
Minor comments, otherwise LGTM

kyouma committed:
Minor updates for doc/whats_new/v0.24.rst

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

kyouma committed:
Minor updates for doc/whats_new/v0.24.rst

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

kyouma committed:
Minor update for comments in sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

glemaitre committed:
nitpicks

glemaitre committed:
PEP8

glemaitre reviewed (approved):
LGTM. I will merge when the CIs will turn green.

glemaitre committed:
Merge remote-tracking branch 'origin/master' into pr/kyouma/18405
